fix: NetworkTransform teleport interpolating from last position on non-authoritative side [NCCBUG-166]

NoelStephensUnity opened the pull request:
This fixes the issue with NetworkTransform and Teleporting as well as can be considered a continuation of #2102 (_this branch started from the fix/networktransform-Bitset-not-being-properly-reset branch_)
This includes several edge case scenarios where non-authoritative instances were still using states that were no longer being set (due to bitset fix). This also cleans up how change in local space is handled internally.

[NCCBUG-166](https://jira.unity3d.com/browse/NCCBUG-166)
[MTT-3458](https://jira.unity3d.com/browse/MTT-3458)
[MTT-4275](https://jira.unity3d.com/browse/MTT-4275)
This also has a portion of the fix from user [xyxwangkai](https://github.com/xyxwangkai)'s [PR-2100](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/2100)

## Changelog
- Fixed: Issue with the internal `NetworkTransformState.m_Bitset` flag not getting cleared upon the next tick advancement.
- Fixed: Interpolation issue with `NetworkTransform.Teleport`.
- Fixed: Issue where the authoritative side was interpolating its transform.

## Testing and Documentation
- Includes updates to existing NetworkTransform integration tests.
- Includes new NetworkTransform integration tests.

### Notes:
While this PR does introduce a more accurate client authoritative model, it still preserves the server authoritative "client directed state" functionality prior to these fixes and updates.  A user can still derive from `NetworkTransform`, not override the `NetworkTransform.OnIsServerAuthoritative` method, and still use the `NetworkTransform.TryCommitTransformToServer` and/or the `NetworkTransform.SetState` methods when an owner is not authoritative but wants to be in control of the state.  This will still add a half RTT penalty for the RPC to make it to the authority before the state is updated (as it did before).  When operating in client authoritative mode, the server can still use both methods as well to preserve the server being the final authority.

The majority of the files in this PR are for manual testing purposes.  This PR can be broken into three parts:
1. The fixes and associated integration tests (4 files under com.unity.netcode.gameobjects)
2. The fix for the memory leak error that has been happening randomly in the editor test: NetworkTimeTests.
3. The manual test updates and additions (24 files under testproject)


The manual tests were left in this branch so anyone reviewing can visually/manually test the teleporting (Samples Menu -> Teleporting) as well as the rest of the scene loading tests to verify everything looks the same (and to run comparison and performance tests against develop branch).

NoelStephensUnity committed:
fix

NetworkTransform was setting the value of m_Bitset but not resetting the individual bits if there was no deltas.  This would result in a gradually increasing message size over time when updating the NetworkTransform.

NoelStephensUnity committed:
update

migrating the assignment back down below the CanCommitToTransform

NoelStephensUnity committed:
fix

This resolves the issue where clients were applying every value marked to be synchronized even if the update didn't include specific axis values.

NoelStephensUnity committed:
test

This includes some fixes to prefabs that hadn't been updated to the more recent changes in NetworkTransform that limited the minimum threshold (i.e. they were set to 0 which causes issues)

NoelStephensUnity committed:
fix

Since we are now no longer sending the entire transform values, this addition prevents the non-authoritative side from making changes to the transform.

NoelStephensUnity committed:
style

removing whitespace

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-Bitset-not-being-properly-reset

NoelStephensUnity committed:
update

Adding some test adjustments and including rotation to interpolation measurements.

NoelStephensUnity committed:
style

removing an unused namespace.

NoelStephensUnity committed:
style

fixing issue with reference to UnityEngine's Object after I removed the system namespace.

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-Bitset-not-being-properly-reset

NoelStephensUnity committed:
fix

This includes the adjustments for client authoritative mode to work without sending RPCs.

NoelStephensUnity committed:
fix

This update fixes the issue where the authoritative side was interpolating when interpolate is enabled.
Now, the authoritative side no longer interpolates.

NoelStephensUnity committed:
manual test updates

These changes make the player client authoritative within the Samples Menu->Additive Scene Loading manual test/sample.

NoelStephensUnity committed:
update

reverting the change of m_Bitset to an internal.

NoelStephensUnity committed:
test update

updating test to account for reverting back to the private m_Bitset.

NoelStephensUnity committed:
update

Migrated the code that updates interpolators with the current value if the axis is not set into the AddInterpolatedState method as opposed to it being in the ApplyInterpolatedNetworkStateToTransform

NoelStephensUnity committed:
fix

First pass fix for teleporting.

NoelStephensUnity committed:
test manual

Manual test adjustments (still working on player motion controller from client side)

NoelStephensUnity committed:
manual test

Adjustments for the teleporting sample/manual test.
Still working on the new player cube motion model.

NoelStephensUnity committed:
test manual

Improved motion model (still could use some more work for client to server side).
Now packing data sent from client to host/server.
Added UI text that displays whether the instance is a server, host, or client.

NoelStephensUnity committed:
update

removing debug code

NoelStephensUnity committed:
fix

This fixes the last issues with teleporting and the updated player motion model.

NoelStephensUnity committed:
fix

Cleaning up how we handle transitioning to and from global and local space.
Minor adjustment to the validation test to assure the range of error being tested on the client side is the same as the delta position threshold range.

NoelStephensUnity committed:
update

Removing the NetworkTransformState Position, Rotation, and Scale since we are only sending the values that changed (i.e. if the x-axis of position was the only thing that was set and NetworkTransformState.Position was used, then the Y and Z axial values would be invalid).

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-Bitset-not-being-properly-reset

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-teleport

NoelStephensUnity committed:
fix

removing code block no longer needed.

NoelStephensUnity committed:
style

removing a comment that was no longer needed.

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-teleport

NoelStephensUnity committed:
update

applying suggested changes

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-Bitset-not-being-properly-reset

NoelStephensUnity committed:
Merge branch 'fix/networktransform-Bitset-not-being-properly-reset' into fix/networktransform-teleport

NoelStephensUnity commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -182,39 +182,6 @@ internal bool IsTeleportingNextFrame
             internal float ScaleX, ScaleY, ScaleZ;
             internal double SentTime;
 
-            internal Vector3 Position
```
This was removed due to the m_Bitset changes.  Since axial values could potentially be invalid (i.e. 0.0f by default when deserialized), these methods were removed to avoid any future changes from using them.  The individual float properties for all axis are used instead.

NoelStephensUnity committed:
update

replacing the GetReplicatedNetworkState() method with a getter ReplicatedNetworkState property.

NoelStephensUnity committed:
Merge branch 'fix/networktransform-Bitset-not-being-properly-reset' into fix/networktransform-teleport

NoelStephensUnity committed:
update

This is a new approach:
- Make in-between state updates additive
- Clear our state's Bitset value and prepare to capture any deltas for the next network tick period.

NoelStephensUnity committed:
update

Realized since we are back to additive (per network tick), we can now remove the non-authoritative interpolation work around for not having all values that changed within the associated NetworkTransformState's network tick period.

NoelStephensUnity committed:
update

Disposing the NetworkVariables when the NetworkTransform is destroyed.

SamuelBellomo commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -182,37 +182,16 @@ internal bool IsTeleportingNextFrame
-                    ScaleZ = value.z;
-                }
+                var localSpace = (ushort)(m_Bitset & (1 << k_InLocalSpaceBit));
+                m_Bitset = (ushort)(m_Bitset & localSpace);
```
you can do &=

SamuelBellomo commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -182,37 +182,16 @@ internal bool IsTeleportingNextFrame
-                    ScaleY = value.y;
-                    ScaleZ = value.z;
-                }
+                var localSpace = (ushort)(m_Bitset & (1 << k_InLocalSpaceBit));
```
bitset is for dirtying certain axis and fields in the transform's state. Why are we keeping localSpace always dirty?

NoelStephensUnity replied:
We are preserving local space settings there.
bitset axial values are now reset every tick, with the exception of the local space setting which applies for all ticks until it is changed by authority.

SamuelBellomo replied:
why are we keeping local space settings for all ticks?

NoelStephensUnity replied:
Because that is only changed by authority on the authority side.  Consider the following:

- First 100 ticks it is in localspace
  - We want to keep that setting in the state
- Tick 101 authority changes it to globalspace
  - We want to keep that setting in the state too.

It just preserves localspace if it is set...that is all.

SamuelBellomo replied:
oh wait, I think I forgot. All others bits are for dirtying, this bit is the actual value, is it? (might want to add a quick comment on this, I forgot about this after almost a year not touching this code base)

NoelStephensUnity replied:
Yep!
The rest are ok to reset when preparing for the next tick, but we have to preserve the local space flag as that is only changed when the NetworkTransform.InLocalSpace is changed on the authoritative side.

SamuelBellomo commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -336,14 +316,15 @@ internal Vector3 Scale
-        /// Note: can be changed during runtime.
         /// </summary>
+        /// <remarks>
+        /// This should only be changed by the authoritative side during runtime. Non-authoritative changes
```
Not sure I understand this comment, interpolation should only be applied on the ghost side, the authority never interpolates.

NoelStephensUnity replied:
Woops!
That comment was supposed to be for InLocalSpace!  
DOH! >.<

SamuelBellomo commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -352,6 +333,7 @@ internal Vector3 Scale
         /// If using different values, please use RPCs to write to the server. Netcode doesn't support client side network variable writing
         /// </summary>
         // This is public to make sure that users don't depend on this IsClient && IsOwner check in their code. If this logic changes in the future, we can make it invisible here
+        // TODO: With recent updates 08-2022, this should be changed to a private set.
```
why not change it now?

SamuelBellomo commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -352,6 +333,7 @@ internal Vector3 Scale
         /// If using different values, please use RPCs to write to the server. Netcode doesn't support client side network variable writing
         /// </summary>
         // This is public to make sure that users don't depend on this IsClient && IsOwner check in their code. If this logic changes in the future, we can make it invisible here
+        // TODO: With recent updates 08-2022, this should be changed to a private set.
```
Is there a ticket or an issue to track this if we ever create a breaking change version?

NoelStephensUnity replied:
Not yet, I am still drafting (but please do continue to review, poke, prod, and challenge any changes... the more eyes the better IMO).  
This is something I am going to put in a list (will send you a link to the document)

SamuelBellomo commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -366,53 +348,112 @@ internal Vector3 Scale
+        /// One for server authoritative and one for "client/owner" authoritative.
+        /// </summary>
+        private readonly NetworkTransformVariable m_ReplicatedNetworkStateServer = new NetworkTransformVariable(new NetworkTransformState(), NetworkVariableReadPermission.Everyone, NetworkVariableWritePermission.Server);
+        private readonly NetworkTransformVariable m_ReplicatedNetworkStateOwner = new NetworkTransformVariable(new NetworkTransformState(), NetworkVariableReadPermission.Everyone, NetworkVariableWritePermission.Owner);
```
Since owner driven transform will be done by overriding NetworkTransform, shouldn't this be done by users when they override it? See PR here https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/1873

NoelStephensUnity replied:
We need to define two NetworkVariables as they have to be instantiated when declared (last I checked) in order to work properly.

SamuelBellomo replied:
My comment above with user defined ClientNetworkTransform would still work, as the child class would declare it the same way. I guess yours works as well, since you have that wrapper netvar that abstracts it.

SamuelBellomo commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -366,53 +348,112 @@ internal Vector3 Scale
-        private BufferedLinearInterpolator<float> m_PositionXInterpolator; // = new BufferedLinearInterpolatorFloat();
-        private BufferedLinearInterpolator<float> m_PositionYInterpolator; // = new BufferedLinearInterpolatorFloat();
-        private BufferedLinearInterpolator<float> m_PositionZInterpolator; // = new BufferedLinearInterpolatorFloat();
-        private BufferedLinearInterpolator<Quaternion> m_RotationInterpolator; // = new BufferedLinearInterpolatorQuaternion(); // rotation is a single Quaternion since each euler axis will affect the quaternion's final value
```
I'd still keep the comment to explain why this one is different from the others

NoelStephensUnity replied:
My bad... it got deleted be accident.  Good catch!

SamuelBellomo commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -366,53 +348,112 @@ internal Vector3 Scale
-        /// Tries updating the server authoritative transform, only if allowed.
-        /// If this called server side, this will commit directly.
-        /// If no update is needed, nothing will be sent. This method should still be called every update, it'll self manage when it should and shouldn't send
+        /// Refer to <see cref="TryCommitTransform">, this is a legacy method.
```
Could you use the Obsolete attribute?
https://docs.microsoft.com/en-us/dotnet/api/system.obsoleteattribute?view=net-6.0

JesseOlmer replied:
The Obsolete tag is a breaking change as it causes either a compile error, or a warning, for anyone using the method. That means introducing it requires a Major version revision. It's effectively useless in the Unity ecosystem unless it's paired with an ApiUpdater, unfortunately.

SamuelBellomo commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -698,53 +733,105 @@ private void ApplyInterpolatedNetworkStateToTransform(NetworkTransformState netw
+                // we should clear our float interpolators
+                foreach (var interpolator in m_AllFloatInterpolators)
+                {
+                    interpolator.Clear();
```
Did you test this with artificial packet drop and latency spikes? If we're clearing the buffered interpolator each time we're receiving state, this effectively makes the interpolator a simple lerper, it removes all buffering.

NoelStephensUnity replied:
This only happens when authority changes InLocalSpace or we are teleporting.

NoelStephensUnity committed:
update

Suggestion updates and comment fixes.

SamuelBellomo commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -844,35 +946,82 @@ private void Awake()
             }
         }
 
+        /// <summary>
```
I usually try to keep most of my properties at the top of the class together, dunno if there's a convention for this in the SDK team

NoelStephensUnity replied:
Yeah, I might be removing these... with the newer fix in place (improved over the Bitset branch), I am backing out other changes (additions) that are no longer needed now that we are additively updating the state during the current network tick up until the current state is sent.  Then it resets to prepare for capturing the transform state for the next tick.

If I end up keep these then they will definitely get migrated up to the rest of the properties. 👍

SamuelBellomo commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -889,15 +1038,33 @@ public override void OnLostOwnership()
         private void Initialize()
         {
-            ResetInterpolatedStateToCurrentAuthoritativeState(); // useful for late joining
+            if (!IsSpawned)
```
wait, how can this happen? If this is called in OnNetworkSpawn, lost and gained ownership, this should never happen, no? Should this throw in that case? (better fail loud and early than silently?)

NoelStephensUnity replied:
This can happen here:
[NetworkObject.InvokeBehaviourNetworkSpawn](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/e138a9c5612e02acf90bfcfbd381aab9ab5b94b3/com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs#L809)

SamuelBellomo replied:
Wait so you can have an ownership callback while not spawned? This seems like a bug no? Shouldn't Ownership callbacks be safe spaces and always online? Does it make sense, from a user's perspective, to have an ownership change while offline?

NoelStephensUnity replied:
When spawning, you need to determine ownership before you spawn so you know who the owner is when OnNetworkSpawn is invoked.

NoelStephensUnity replied:
It isn't really "not spawned" but it isn't really "spawned"... it is the quasi state while in the middle of spawning.  It is just an order of operations thing... users want to know who the owner is when OnNetworkSpawn is invoked...so before that is invoked it has to set ownership before invoking OnNetworkSpawn when in the middle of spawning.

Make sense?

SamuelBellomo replied:
I agree that internal ownership book keeping should happen before spawn, but should the user side callback be called at that time too? Couldn't internal ownership changes be made, spawn called, then gained ownership callback called?

NoelStephensUnity replied:
In [NetworkSpawnManager.SpawnNetworkObjectLocallyCommon](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/e138a9c5612e02acf90bfcfbd381aab9ab5b94b3/com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs#L552) both client and server invoke this method at the same place.  

It is a global "order of operations" thing that we can't change as it would change core functionality.
This really isn't that big of a deal if you think about it:

During Spawn:
Ownership initialization is skipped, but OnNetworkSpawn initialization is not.  Since they both call the same method that invokes the same logic, it "just works". 👍 

During Ownership Changes When Spawned:
Ownership initialization is not skipped and since it is spawned we don't worry about OnNetworkSpawn.

SamuelBellomo commented on com.unity.netcode.gameobjects/Tests/Runtime/NetworkTransform/NetworkTransformTests.cs:
```diff
@@ -118,38 +117,38 @@ public IEnumerator TestAuthoritativeTransformChangeOneAtATime([Values] bool test
-            Assert.LessOrEqual(Math.Abs(40 - otherSideNetworkTransform.transform.rotation.eulerAngles.y), approximation, $"wrong rotation on ghost on y, got {otherSideNetworkTransform.transform.rotation.eulerAngles.y}");
-            Assert.LessOrEqual(Math.Abs(35 - otherSideNetworkTransform.transform.rotation.eulerAngles.z), approximation, $"wrong rotation on ghost on z, got {otherSideNetworkTransform.transform.rotation.eulerAngles.z}");
+            //// approximation needed here since eulerAngles isn't super precise.
+            //Assert.LessOrEqual(Math.Abs(45 - otherSideNetworkTransform.transform.rotation.eulerAngles.x), approximation, $"wrong rotation on ghost on x, got {otherSideNetworkTransform.transform.rotation.eulerAngles.x}");
```
if those are covered above, should we remove those comments?

NoelStephensUnity replied:
Indeed we should.  👍 
Once I have finalized the changes, I will be adding additional integration tests and will clean this area up for sure.

SamuelBellomo commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -182,37 +182,16 @@ internal bool IsTeleportingNextFrame
-                    ScaleY = value.y;
-                    ScaleZ = value.z;
-                }
+                var localSpace = (ushort)(m_Bitset & (1 << k_InLocalSpaceBit));
```
```suggestion
                var localSpace = (ushort)(m_Bitset & (1 << k_InLocalSpaceBit)); // since we store the actual value (and not a dirty bit) here, we need to preserve it
```

NoelStephensUnity replied:
Good call.

NoelStephensUnity committed:
style

adding Sam's comment.

NoelStephensUnity committed:
style

updated comment to include what we are preserving.

NoelStephensUnity committed:
update

Removing the ApplyTransformValues and SetTransformValues as they are no longer needed.
Also, removing the associated properties.

NoelStephensUnity committed:
Test Manual

Updates to manual test related assets and scripts.
Removed the "make a pooled object invisible to hide the teleport interpolation issue" hack that would disable the render mesh when despawned and enable it shortly after it is spawned.
Increased tick rate to 32 in the additive scene loading test.
Fixing white space issues.

NoelStephensUnity committed:
update

Removing the unused serverTime addition as well.

NoelStephensUnity committed:
update

Renamed ApplyInterpolatedNetworkStateToTransform to ApplyAuthoritativeState.
(It didn't make sense if interpolation is disabled)

NoelStephensUnity committed:
tests update

This includes an updated NetworkTransformTests that tests both authority types while running in Server and Host modes.
This includes some updates to naming for easier identification of what each test does (in test runner and in code).

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-teleport

NoelStephensUnity committed:
Update

Minor adjustment where we don't need to send a local properties over the stack via parameters.

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-teleport

NoelStephensUnity committed:
test

Added interpolate enable/disable values to all tests.
Now, all NetworkTranaformTests are done with interpolation enabled and disabled.

NoelStephensUnity committed:
fix and update

Fixed another issue that was exposed while cleaning up the code base.
Was able to remove even more of the code added previously.
Some clean up as well as making one additional method private as opposed to protected to prevent us from having to release a minor version vs patch.
Also,  ApplyTransformToNetworkState, and ApplyTransformToNetworkStateWithInfo only return if the state is dirty (the rest can be determined from within the state itself).

ApplyLocalNetworkState, used for integration tests, now just returns the NetworkTransformState.

NoelStephensUnity committed:
fix

This fixes the issue with the memory leak message in the editor NetworkTimeTests  (very odd this fixed the issue).

NoelStephensUnity committed:
test manual

Switched over to an owner authoritative NetworkTransform that also includes the playermovement.
Adjusted the PlayerCube NetworkPrefab and minor adjustment to the CollideTeleporter.

NoelStephensUnity committed:
update

updating the changelog

NoelStephensUnity committed:
fix

Found a minor logic issue in ApplyTransformToNetworkStateWithInfo where the state dirty flag could be reset before end of the current network tick.  This could result in state being carried forward into the next network tick.  This resolves that issue.

NoelStephensUnity committed:
update

Minor adjustment to how we handle applying a rotation measurement on the non-authoritative side.
Removed a property not being used outside of a single method and added additional comments to the PlayerMovement component.

NoelStephensUnity committed:
update

This update includes the final changes required in order to support the previous NetworkTransform model where it was always server authoritative and client owners could direct/apply state via internal RPCs.  This also includes the ability for a server to apply/dictate state when running in client owner authority mode.

This includes updates to the NetworkTransform tests that validate all of these updates and increase the amount of coverage for NetworkTransform.

NoelStephensUnity committed:
style

moving new private m_ClientRpcParams and m_ClientIds up to the rest of the property declarations.
Adjusting comment

NoelStephensUnity committed:
test manual

Minor adjustments for the random player movers.
Applying the server authoritative player prefab back to be the default for the additive scene sample (this makes it easier to compare to main branch).

NoelStephensUnity committed:
style

removing a no longer needed comment.
adding note about future clean up for NetworkTransform.

NoelStephensUnity committed:
style

fixed spelling issue

NoelStephensUnity committed:
style

removing left over comment and spelling

NoelStephensUnity committed:
style

removing legacy/unused namespaces

NoelStephensUnity committed:
update

realized our CommitLocallyAndReplicate method was serving no purpose at this point.
added the assignment of NetworkTransformState to the ReplicatedNetworkState within TryCommit and removed CommitLocallyAndReplicate.

NoelStephensUnity committed:
update

removing legacy check for TryCommitTransformToServer that is no longer needed.

NoelStephensUnity committed:
update

removing m_Transform (no purpose for it)
Fixing minor grammar issue with a comment

NoelStephensUnity committed:
test manual update

removing additional IsOwner check.

NoelStephensUnity committed:
revert

reverting the changes I made to the editor test NetworkTimeTests as I just had Yamato crash again in a different area in NetworkTimeTests with the same memory leak error.

NoelStephensUnity committed:
style

replacing LF/CR

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-teleport

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-teleport

ShadauxCat commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -366,62 +383,107 @@ internal Vector3 Scale
+        /// We have two internal NetworkVariables.
+        /// One for server authoritative and one for "client/owner" authoritative.
+        /// </summary>
+        private readonly NetworkVariable<NetworkTransformState> m_ReplicatedNetworkStateServer = new NetworkVariable<NetworkTransformState>(new NetworkTransformState(), NetworkVariableReadPermission.Everyone, NetworkVariableWritePermission.Server);
```
Since the difference between these two is only a matter of what parameters are passed to the constructor, and from what I can tell only one of them is ever used, couldn't these just be collapsed to just one that's initialized in OnNetworkSpawn? Granted as it's currently implemented we have no way to react to a change from server- to owner-authoritative... but that seems to be a problem in the current implementation, too, since changing that would suddenly change it to using the other one that's probably gonna be all 0s?

NoelStephensUnity replied:
You can't initialize NetworkVariables in OnNetworkSpawn.  By then it is too late... they have to be initialized either when they are declared or in Awake.  Since user code could declare an Awake within a child derived class, this would break the initialization since the NetworkTransform.Awake would not be invoked.  They could also change a property value returned in the IsServerAuthoritative in their Start method.

It is the safest way to do this for the patch.

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-teleport

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-teleport

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-teleport

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-teleport

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-teleport

ShadauxCat commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -432,163 +494,141 @@ void Send(NetworkTransformState stateToSend)
             }
             else if (!m_HasSentLastValue && m_CachedNetworkManager.LocalTime.Tick >= m_LastSentTick + 1) // check for state.IsDirty since update can happen more than once per tick. No need for client, RPCs will just queue up
             {
+                m_LastSentState.IsTeleportingNextFrame = false; // This is required here
```
"This is required here" - can you elaborate in the comment, so when we come back to this later, we not only know that it's required, but why?

NoelStephensUnity replied:
Heh... yeah... I was wondering if I needed to elaborate on that or not.
Basically, that portion of the code is just resetting the IsTeleportingNextFrame because of the m_LastSentState as it is used (based on the legacy notes above it) to help clients extrapolate to the last position (i.e. making them motionless) if there are no more deltas to be sent.
This is called out in the document and I think this "else if" segment of code can be removed completely (it requires us to migrate this functionality over to the non-authoritative side).

ShadauxCat replied:
I meant, can you add this to the comment in the code, so that if someone comes here to refactor this later (and doesn't know about the document) they can understand the reasons that has to be right there specifically?

ShadauxCat commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -432,163 +494,141 @@ void Send(NetworkTransformState stateToSend)
             m_PositionZInterpolator.ResetTo(m_LocalAuthoritativeNetworkState.PositionZ, serverTime);
 
-            m_RotationInterpolator.ResetTo(Quaternion.Euler(m_LocalAuthoritativeNetworkState.Rotation), serverTime);
+            m_RotationInterpolator.ResetTo(Quaternion.Euler(m_LocalAuthoritativeNetworkState.RotAngleX, m_LocalAuthoritativeNetworkState.RotAngleY, m_LocalAuthoritativeNetworkState.RotAngleZ), serverTime);
```
I don't know if this really matters that much, but when I was doing FastBufferReader/Writer I had trouble writing tests when I was doing quaternions this way, because quaternion -> euler -> quaternion is not a lossless process. The quaternion would always end up different so I changed it to just send the quaternion directly instead of converting it to euler angles because that was the only way I could get it deterministic enough to test.

This isn't really related to your changes but felt worth mentioning. I don't know enough about quaternions to know whether that loss in the conversion might cause any issues.

NoelStephensUnity replied:
You do lose resolution but it won't be more than the minimum threshold value to detect change.  
So, we are not 100% lossless but relative to the resolution being sent it works out.

ShadauxCat commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -432,163 +494,141 @@ void Send(NetworkTransformState stateToSend)
 
         /// <summary>
-        /// Will apply the transform to the LocalAuthoritativeNetworkState and get detailed isDirty information returned.
+        /// Used for integration testing:
```
Any time I see a comment like this, I wonder if this method should be here. Could the tests replace NetworkTransform with a subclass that adds this method?

NoelStephensUnity replied:
That was there prior to the update.  I just adjusted the comment to call it out for that purpose.
I added that to the list of things we can continue to improve upon for NetworkTransform.

ShadauxCat replied:
👍 Recording a todo for later is good enough for now.

ShadauxCat commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -432,163 +494,141 @@ void Send(NetworkTransformState stateToSend)
 
-        private (bool isDirty, bool isPositionDirty, bool isRotationDirty, bool isScaleDirty) ApplyTransformToNetworkStateWithInfo(ref NetworkTransformState networkState, double dirtyTime, Transform transformToUse)
+        /// <summary>
+        /// Used for integration testing
```
See above comment.

NoelStephensUnity replied:
See above reply.  :)

ShadauxCat commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -602,149 +642,221 @@ internal bool ApplyTransformToNetworkState(ref NetworkTransformState networkStat
 
-        private void AddInterpolatedState(NetworkTransformState newState, bool reset = false)
+        /// <summary>
+        /// Only non-authoritative instances should invoke this
```
Should there be a check in the method to enforce this comment?

NoelStephensUnity replied:
Really, there is only one method that calls that:
```
        private void OnNetworkStateChanged(NetworkTransformState oldState, NetworkTransformState newState)
        {
            if (!NetworkObject.IsSpawned)
            {
                return;
            }

            if (CanCommitToTransform)
            {
                // we're the authority, we ignore incoming changes
                return;
            }

            if (Interpolate)
            {
                AddInterpolatedState(newState);
            }
        }
```
We really should migrate that code into the OnNetworkStateChanged method.  Was trying to reduce the complexity of the PR (already removed several methods).
It is added to the list.  👍

ShadauxCat replied:
Oh, right, it's private, not protected. Ok, withdrawn.

ShadauxCat commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -698,53 +733,105 @@ private void ApplyInterpolatedNetworkStateToTransform(NetworkTransformState netw
+                // we should clear our float interpolators
+                foreach (var interpolator in m_AllFloatInterpolators)
+                {
+                    interpolator.Clear();
```
Seems like a comment to help reviewers understand the code should be in the code. 😏

NoelStephensUnity replied:
Heh...yeah... well it diverges from the original file so much I felt a note there was helpful.

ShadauxCat commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -844,35 +975,39 @@ private void Awake()
+            Initialize();
+
+            // This assures the initial spawning of the object synchronizes all connected clients
+            // with the current transform values. This should not be placed within Initialize since
```
Thanks for this comment, warnings like this against moving code around are really nice.

ShadauxCat commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -844,35 +975,39 @@ private void Awake()
-            // crucial we do this to reset the interpolators so that recycled objects when using a pool will
-            //  not have leftover interpolator state from the previous object
-            Initialize();
+                // Force the state update to be sent
```
I think it's interesting here that the comment says "Force" but the method says "Try". I'm sure it does what the comment says it does, but it's an interesting contradiction of language.

NoelStephensUnity replied:
Yeah... I think we still have some names to update in NetworkTransform.

NoelStephensUnity committed:
Style

Improving comments
